Inspecting the standard library

I would like a way to automatically find all the types in `std`, but naively pointing `cargo public-api` at `std` fails because it can't find `core` and other dependencies. 

```console
$ cargo public-api --manifest-path $(rustc --print sysroot)/lib/rustlib/src/rust/library/std/Cargo.toml
Error: Failed to build rustdoc JSON. Stderr:     Checking compiler_builtins v0.1.78
    Checking libc v0.2.127
   Compiling unwind v0.0.0 (/home/michael/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/unwind)
error[E0433]: failed to resolve: could not find `prelude` in `core`

error[E0432]: unresolved import `core::ops`
 --> /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.78/src/float/mod.rs:1:5
  |
1 | use core::ops;
  |     ^^^^^^^^^ no `ops` in the root

error[E0432]: unresolved import `core::intrinsics`
  --> /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.78/src/mem/mod.rs:11:11
   |
11 | use core::intrinsics::{atomic_load_unordered, atomic_store_unordered, exact_div};
   |           ^^^^^^^^^^ could not find `intrinsics` in `core`

...
```

Presumably `rustc` does something funky with the sysroot that `cargo public-api` doesn't).

Do you have any suggestions for how I might be able to achieve this?

--- a/cargo-public-api/src/main.rs
+++ b/cargo-public-api/src/main.rs
@@ -76,6 +76,19 @@ pub struct Args {
     #[clap(long, min_values = 2, max_values = 2)]
     diff_rustdoc_json: Option<Vec<String>>,
 
+    /// Usage: --rustdoc-json <RUSTDOC_JSON_PATH>
+    ///
+    /// List the public API based on the given rustdoc JSON file. Try for example
+    ///
+    ///     rustup component add rust-docs-json --toolchain  nightly
+    ///
+    /// and then
+    ///
+    ///     cargo public-api --rustdoc-json ~/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/share/doc/rust/json/std.json
+    ///
+    #[clap(long)]
+    rustdoc_json: Option<String>,
+
     /// Exit with failure if the specified API diff is detected.
     ///
     /// * all = deny added, changed, and removed public items in the API
@@ -167,6 +180,8 @@ fn main_() -> Result<()> {
         print_diff_between_two_commits(&args, commits)?
     } else if let Some(files) = &args.diff_rustdoc_json {
         print_diff_between_two_rustdoc_json_files(&args, files)?
+    } else if let Some(rustdoc_json) = &args.rustdoc_json {
+        print_public_items_from_json(&args, rustdoc_json)?
     } else {
         print_public_items_of_current_commit(&args)?
     };
@@ -258,7 +273,19 @@ fn check_diff(args: &Args, diff: &Option<PublicItemsDiff>) -> Result<()> {
 
 fn print_public_items_of_current_commit(args: &Args) -> Result<PostProcessing> {
     let (public_api, branch_to_restore) = collect_public_api_from_commit(args, None)?;
+    print_public_items(args, &public_api, branch_to_restore)
+}
 
+fn print_public_items_from_json(args: &Args, json_path: &str) -> Result<PostProcessing> {
+    let public_api = public_api_from_rustdoc_json_path(json_path, args)?;
+    print_public_items(args, &public_api, None)
+}
+
+fn print_public_items(
+    args: &Args,
+    public_api: &PublicApi,
+    branch_to_restore: Option<String>,
+) -> Result<PostProcessing> {
     Plain::print_items(&mut stdout(), args, &public_api.items)?;
 
     Ok(PostProcessing {

--- a/cargo-public-api/tests/cargo-public-api-bin-tests.rs
+++ b/cargo-public-api/tests/cargo-public-api-bin-tests.rs
@@ -389,6 +389,17 @@ fn diff_public_items_from_files() {
 }
 
 #[test]
+fn list_public_items_from_json_file() {
+    let json_file = rustdoc_json_path_for_crate("../test-apis/example_api-v0.3.0");
+    let mut cmd = Command::cargo_bin("cargo-public-api").unwrap();
+    cmd.arg("--rustdoc-json");
+    cmd.arg(json_file);
+    cmd.assert()
+        .stdout_or_bless("./tests/expected-output/example_api-v0.3.0.txt")
+        .success();
+}
+
+#[test]
 fn diff_public_items_missing_one_arg() {
     let mut cmd = TestCmd::new();
     cmd.arg("--diff-git-checkouts");

--- a/public-api/Cargo.toml
+++ b/public-api/Cargo.toml
@@ -23,7 +23,7 @@ features = ["unbounded_depth"]
 
 [dependencies.rustdoc-types]
 # path = "/Users/martin/src/rustdoc-types"
-version = "0.17.0"
+version = "0.18.0"
 
 [dev-dependencies]
 anyhow = "1.0.53"

--- a/public-api/src/render.rs
+++ b/public-api/src/render.rs
@@ -176,7 +176,21 @@ impl<'a> RenderingContext<'a> {
                 }
                 output
             }
-            ItemEnum::PrimitiveType(_) => self.render_simple(&["primitive", "type"], &item.path()),
+            ItemEnum::Primitive(primitive) => {
+                // This is hard to write tests for since only Rust `core` is
+                // allowed to define primitives. So you have to test this code
+                // using the pre-built rustdoc JSON for core:
+                //
+                //   rustup component add rust-docs-json --toolchain nightly
+                //   cargo run -- --rustdoc-json ~/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/share/doc/rust/json/core.json
+                let mut output = pub_();
+                output.extend([
+                    Token::kind("type"),
+                    ws!(),
+                    Token::primitive(&primitive.name),
+                ]);
+                output
+            }
         };
 
         tokens.extend(inner_tokens);

